feat: refactor and improve K8s sync provider

Kavindu-Dodan opened the pull request:
## This PR
fixes #434

This is a refactoring and internal improvement for K8s ISync provider.

Improvements include,

- Reduce K8s API load by utilizing Informer cache
- Yet, provide a fallback if cache miss occurs (Note - we rely on K8s Informer for cache refill and consistency)
- Informer now only watches a specific namespace (compared to **\***) - this is a potential performance improvement and a security improvement
- Reduced informer handlers with extracted common logics 
- Unit tests where possible

thisthat reviewed (approved):
I left only minor suggestions to improve logging a bit. I see that parts of the code are not tested, in particular the one with K8s details. For that, I recommend having a look at the fake client [(docs)](https://pkg.go.dev/k8s.io/client-go/dynamic/fake)

toddbaert commented on pkg/sync/kubernetes/kubernetes_sync_test.go:
```diff
@@ -0,0 +1,362 @@
+package kubernetes
```
Thanks for adding the tests. I kept meaning to add an issue to implement these and forgot to do it.

Kavindu-Dodan replied:
Thanks. Yes, I added tests for easily testable parts. However, we still need coverage on K8s-backed logic. These need either mocking or true e2e tests.

toddbaert replied:
We already have the "true" e2e tests in OFO. I think that's enough.

I think though we should improve coverage in this repo as well..

Kavindu-Dodan commented on pkg/sync/kubernetes/kubernetes_sync.go:
```diff
@@ -23,27 +22,67 @@ import (
+		return err
+	}
+
+	clusterClient, err := dynamic.NewForConfig(clusterConfig)
```
changed in the latest commit

Kavindu-Dodan commented on pkg/sync/kubernetes/kubernetes_sync.go:
```diff
@@ -210,48 +195,52 @@ func (k *Sync) notify(ctx context.Context, c chan<- INotify) {
-	if ffObj.APIVersion != fmt.Sprintf("%s/%s", v1alpha1.GroupVersion.Group, v1alpha1.GroupVersion.Version) {
+
+	if ffObj.APIVersion != apiVersion {
 		return errors.New("invalid api version")
```
improved in the latest commit

toddbaert commented:
I tested this in OFO locally (latest version) and it works as expected.

toddbaert commented on pkg/sync/kubernetes/kubernetes_sync.go:
```diff
@@ -91,101 +130,47 @@ func (k *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
-
-	ns, name, err := parseURI(k.URI)
+	// first check the store - avoid overloading API
+	object, exist, err := k.informer.GetStore().GetByKey(k.URI)
```
```suggestion
	rawCfg, exist, err := k.informer.GetStore().GetByKey(k.URI)
```

It doesn't have to be this name, but I think implying this is unparsed JSON or a raw payload would be clearer.

Kavindu-Dodan replied:
resolved with the latest commit

Kavindu-Dodan commented:
> I left only minor suggestions to improve logging a bit. I see that parts of the code are not tested, in particular the one with K8s details. For that, I recommend having a look at the fake client [(docs)](https://pkg.go.dev/k8s.io/client-go/dynamic/fake)

Thank you. I addressed the log and renamed related comments. Regarding testing, K8s backed logic. Even with a fake client, we still not to mock the shared informer interface as we have a fallback. I am not sure how worth these mocks are against real e2e tests which easily test most of the logic.

toddbaert commented:
I agree with @thisthat that we probably need a bit more coverage here (I know that there was no tests here before, so that's not your fault). I think now is the time. If there's any way to mock the k8s data I think we should do some basics here. It's worth another day or so of effort I think.

Maybe @thisthat has some specific recommendations.

Besides that everything here looks good to me and I think this is an improvement.

thisthat commented:
K8s provides great fake clients that can be used to unit-test your informer logic. 
I think a good starting point to see an example is this test file in Keptn: https://github.com/keptn/lifecycle-toolkit/blob/91e57cadba32fce6d873bc480408f90bcb8964d9/metrics-operator/cmd/metrics/adapter/provider/provider_test.go

toddbaert commented:
@Kavindu-Dodan we can create a new issue for additional testing. I think at a minimum we should get the coverage diff here to 0% with some meaningful tests unless it's really impossible. This isn't a place where we can afford to lose coverage.

Kavindu-Dodan commented:
> @Kavindu-Dodan we can create a new issue for additional testing. I think at a minimum we should get the coverage diff here to 0% with some meaningful tests unless it's really impossible. This isn't a place where we can afford to lose coverage.

It is not impossible. It's time to deliver. Soon there will be merged conflicts with upcoming PRs and given that we constantly change API contracts, the efforts can go wasted.

Besides, I introduced new tests and simply touched code with the refactoring, reducing complexity. Even with new tests, for some reason, we see a reduction in coverage. To be honest I do not know how this happened.

Kavindu-Dodan commented:
> > @Kavindu-Dodan we can create a new issue for additional testing. I think at a minimum we should get the coverage diff here to 0% with some meaningful tests unless it's really impossible. This isn't a place where we can afford to lose coverage.
> 
> It is not impossible. It's time to deliver. Soon there will be merged conflicts with upcoming PRs and given that we constantly change API contracts, the efforts can go wasted.
> 
> Besides, I introduced new tests and simply touched code with the refactoring, reducing complexity. Even with new tests, for some reason, we see a reduction in coverage. To be honest I do not know how this happened.

To prove my point, here is a PR with the change we need - https://github.com/open-feature/flagd/pull/451

This will work and since I am not refactoring, code coverage won't complain.

Kavindu-Dodan commented:
@thisthat @toddbaert  I added more tests however the coverage is below the previous value due to the reason I explained [1]

Please have another round of review :) 

[1] https://github.com/open-feature/flagd/pull/443#issuecomment-1450710014

james-milligan commented on pkg/sync/kubernetes/kubernetes_sync.go:
```diff
@@ -273,22 +281,48 @@ func updateFuncHandler(oldObj interface{}, newObj interface{}, object client.Obj
 
-func deleteFuncHandler(obj interface{}, object client.ObjectKey, c chan<- INotify) error {
+// toFFCfg attempts to covert unstructured payload to configurations
+func toFFCfg(object interface{}) (*v1alpha1.FeatureFlagConfiguration, error) {
```
if I add the following test to the `Test_toFFCfg` function:
```
{
			name: "Wrong resource type",
			input: toUnstructured(t, corev1.Pod{
				TypeMeta: v1.TypeMeta{
					Kind: "Pod",
				},
			}),
			wantErr: true,
		},
```
the test fails as there is no error returned, this is expected behaviour as the marshalling would not fail, however, IMO it should. We should add an if statement to ensure that the apiversion and kind match at this point, of course this would require the k8s api to return us the wrong object, however I think its worthwhile covering this case as it currently passes with any object

Kavindu-Dodan replied:
Yes, noticed this while testing the function. However, I will keep this as it is. The reason is that if we try to validate, we have to enforce validation criteria. API version is only one such validation. So it is best to allow the caller to perform such validation based on its requirement and have this common logic of parsing here.

james-milligan commented on pkg/sync/kubernetes/kubernetes_sync.go:
```diff
@@ -220,48 +224,52 @@ func (k *Sync) notify(ctx context.Context, c chan<- INotify) {
-		return errors.New("invalid api version")
+
+	if ffOldObj.APIVersion != apiVersion {
+		return fmt.Errorf("invalid api version %s, expected %s", ffOldObj.APIVersion, apiVersion)
```
can duplicate `"Simple error - API version mismatch"` test and change the api version in `validFFCfgOld` to cover this line

james-milligan commented on pkg/sync/kubernetes/kubernetes_sync.go:
```diff
@@ -2,16 +2,15 @@ package kubernetes
 	"fmt"
 	"os"
 	"strings"
+	parallel "sync"
```
nitpick but when the flagd sync package and `"sync"` are being imported in the same package it  has previously been imported as "msync", might cause confusion to change the name in this package, maybe?

Kavindu-Dodan commented on pkg/sync/kubernetes/kubernetes_sync.go:
```diff
@@ -101,101 +159,47 @@ func (k *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
 	if err != nil {
-		k.Logger.Error(err.Error())
-		return "{}", nil
+		return "", err
```
Return value consists of (flagConfig, error). And if there is an error, return the value of flagConfig must be ignored (error handling). This is why the simple empty string return here.

toddbaert commented:
@Kavindu-Dodan @thisthat I created https://github.com/open-feature/flagd/issues/460 for testing.
